// protractor-core/conf/elancelive_conf.js
/**
 * @name elancelive_conf.js
 * @description Standard conf to be use for automation for live site.
 *
 */

var HtmlReporter = require('protractor-html-screenshot-reporter');
var baseConf = require('./base_conf.js');
var baseConfig = new baseConf();

exports.config = {

  seleniumAddress: 'http://localhost:4444/wd/hub',

  capabilities: {
    'browserName': 'chrome',
    'chromeOptions': {
      'args': [
        '--keep-alive-for-test',
        ]
    }
  },

  specs: [
    '../specs/WP-lam_spec.js'
  ],

  params: {
    report: 'result',
    globalTimeout: baseConfig.timeout.pageLoad + 5000,
    waitForAngular: {
      angularFlag: true,  //Only use false if timeout/interval for polling is failing.
      angularDelay: baseConfig.timeout.angularDelay + 5000 //this only goes into affect if angularFlag=false
    }
  },

  allScriptsTimeout: baseConfig.timeout.allScriptsTimeout,

  onPrepare: function () {
    browser.manage().timeouts().pageLoadTimeout(baseConfig.timeout.pageLoad + 5000);
    browser.manage().timeouts().implicitlyWait(baseConfig.timeout.implicitWait + 5000);
    browser.driver.manage().window().setSize(1000,1000);
    jasmine.getEnv().addReporter(new HtmlReporter({
      baseDirectory: browser.params.report,
      docTitle: 'Reports',
      takeScreenShotsOnlyForFailedSpecs: true
    }));
  },

// Options to be passed to Jasmine-node.
  jasmineNodeOpts: {
    showColors: true,
    defaultTimeoutInterval: baseConfig.timeout.defaultTimeoutInterval
  }

}
;
