// protractor-core/conf/env_conf.js
/**
 * This file contains environmental data sets for testing environment: dev/stage/prod.
 * Before access the data, you need to add one commandline prameter. For example, --params.testenv=dev, this will get the data set for the dev environment
 * Default value for this parameter must be defined in your $project_conf.js and the default value is "stage"
 *
 * Sample code to access environmental variables:
 *
 *   var envConf = require('../node_modules/protractor-core/conf/env_conf.js'), env = new envConf();
 *
 *   console.log(env.data.roomsEndPointsUrl);
 */


var envConf = function(){
  this.envDataSet = [
    {
      envConfig: "prod",
      host: 'https://www.upwork.com',
      user: {
        em: 'em130131043801',
        ic: 'ic130130011901'
      },
      agora: {
        app: 'http://front.prod.agora.odesk.com/',
        profile: 'http://front.prod.agora.odesk.com:8204/pub/profiles', // DS = 8206 instead of AS
        job: 'http://front.prod.agora.odesk.com:8184/pub/jobs', // DS = 8186
        category: 'http://front.prod.agora.odesk.com:8440/classifications/default',
        skills: 'http://front.prod.agora.odesk.com:8308/skills',
        logs: 'http://logs.prod.agora.odesk.com'
      },
      match: {
        profile: 'http://match.prod.svc.odesk.com/search/b/v1/profiles',
        job: 'http://match.prod.svc.odesk.com/search/b/v3/jobs',
        tier: 'http://match.prod.svc.odesk.com/dpodw/serve/latest/contractor_predicted_tier/'
      },
      solr: {
        profile: 'http://172.27.65.98:9898/profiles/select',
        job: 'http://172.27.65.231:10300/openings/select'
      },
      odr: {
        profile: 'http://odr.prod.sv4.odesk.com/odr/profile/f/',
        job: 'http://odr.prod.sv4.odesk.com/odr/opening/f/'
      },
      ods: {
        skill: 'http://ods.prod.sv4.odesk.com:83/ods/agg_skill_detail/pa',
        skills: 'http://ods.prod.sv4.odesk.com:83/ods/skill/list_ex'
      },
      orpc: {
        app: 'https://dsokolovsky:passw0rd@s-int.odesk.com:443/orpc/'
      }
    },
    {
      envConfig: "stage",
      host: 'https://stage.upwork.com',
      user: {
        admin: { login: 'dsokolovsky', password: 'passw0rd'},
        em: 'em15040907514912',
        ic: 'qwertyuiop01'
      },
      agora: {
        app: 'http://front.staging.agora.odesk.com/',
        profile: 'http://front.staging.agora.odesk.com:8204/pub/profiles',
        job: 'http://front.staging.agora.odesk.com:8184/pub/jobs',
        category: 'http://front.staging.agora.odesk.com:8440/classifications/default',
        skills: 'http://front.staging.agora.odesk.com:8308/skills',
        registrations: 'http://front.staging.agora.odesk.com:8454/registrations',
        logs: 'http://logs.staging.agora.odesk.com'
      },
      match: {
        profile: 'http://match.staging.svc.odesk.com/search/b/v1/profiles',
        job: 'http://match.staging.svc.odesk.com/search/b/v3/jobs',
        tier: 'http://match.staging.svc.odesk.com/dpodw/serve/latest/contractor_predicted_tier/'
      },
      solr: {
        profile: 'http://172.27.67.219:9898/profiles/select',
        job: 'http://172.27.67.106:10300/openings/select'
      },
      odr: {
        profile: 'http://odr.staging1.sv4.odesk.com/odr/profile/f/',
        job: 'http://odr.staging1.sv4.odesk.com/odr/opening/f/'
      },
      ods: {
        skill: 'http://ods.staging1.sv4.odesk.com:83/ods/agg_skill_detail/pa',
        skills: 'http://ods.staging1.sv4.odesk.com:83/ods/skill/list_ex'
      },
      orpc: {
        app: 'https://dsokolovsky:passw0rd@s-int.upwork.com:443/orpc/',
        url : 'https://stage-web01.odesk.com/rpcapi/index.php',
        uri   : 'http://localhost:82/orpc/',
        authorization : 'Basic ZHNva29sb3Zza3k6cGFzc3cwcmQ=',
        origin  : 'https://stage-web01.odesk.com'
      },
      db: {
        host: 'stg06.odesk.com',
        credentials: {
          'odesk_qa':       {login: 'odesk_qa',       password: 'odesk_qa'},
          'odesk_ba':       {login: 'odesk_ba',       password: 'odesk_ba'},
          'odesk_obo':      {login: 'odesk_obo',      password: 'odesk_obo'},
          'odesk_ua':       {login: 'odesk_ua',       password: 'odesk_ua'},
          'odb2':           {login: 'odb2_user',      password: 'odb2_user'},
          'odesk_projects': {login: 'odesk_projects', password: 'LAJYvLthyhUFaxNPJeuI'},
          'odesk_misc':     {login: 'odesk_misc',     password: 'odesk_misc'}
        }
      },

      roomsEndPointsUrl: 'http://front.staging.agora.odesk.com:8246/rooms',
      personPostEndPointsUrl: 'http://front.staging.agora.odesk.com:8224/persons',
      personEndPointsUrl: 'http://front.staging.agora.odesk.com:8064/persons',
      contactEndPointsUrl: 'http://front.staging.agora.odesk.com:8224/contacts',
      subscriptionEndPointsUrl: 'http://front.staging.agora.odesk.com:8268/subscription',
      pricingEndPointsUrl: 'http://front.staging.agora.odesk.com:8268/pricing',
      connectsEndPointsUrl: 'http://front.staging.agora.odesk.com:8282/connects',
      promotionsEndPointsUrl: 'http://front.staging.agora.odesk.com:8328',
      subscriptionBatchServicesDSEndPoint : 'http://front.staging.agora.odesk.com:8374/current/trigger?',
      orpcNotifyUserVerifyEmail: 'https://dsokolovsky:passw0rd@s-int.upwork.com:443/orpc/NotifyUser::VerifyEmail',
      orpcMailGetMessage: 'https://dsokolovsky:passw0rd@s-int.upwork.com:443/orpc/Mail::GetMessage',
      orpcRegistrationVerifyEmail: 'https://dsokolovsky:passw0rd@s-int.upwork.com:443/orpc/Registration::VerifyEmail',
      orpcQTestsAllocateToMultipleTests: 'https://dsokolovsky:passw0rd@s-int.upwork.com:443/orpc/QTests::AllocateToMultipleTests',
      orpcTeamRESTAdapterAddMemoByCID: 'https://dsokolovsky:passw0rd@s-int.upwork.com:443/orpc/TeamRESTAdapter::AddMemoByCID',
      orpcRegistrationStart: 'https://dsokolovsky:passw0rd@s-int.upwork.com:443/orpc/Registration::Start'

    },
    {
      envConfig: "dev",
      host: 'https://dev03.odesk.com:33051',
      user: {
        admin: { login: 'admin', password: 'strange!'},
        em: 'em14092400220515',
        ic: 'ic15012212025612'
      },
      agora: {
        app: 'http://front.dev.agora.odesk.com/',
        profile: 'http://front.dev.agora.odesk.com:8204/pub/profiles',
        job: 'http://front.dev.agora.odesk.com:8184/pub/jobs',
        category: 'http://front.dev.agora.odesk.com:8440/classifications/default',
        skills: 'http://front.dev.agora.odesk.com:8308/skills',
        logs: 'http://logs.dev.agora.odesk.com'
      },
      match: {
        profile: 'http://match.dev.svc.odesk.com/search/b/v1/profiles',
        job: 'http://match.dev.svc.odesk.com/search/b/v3/jobs',
        tier: 'http://match.dev.svc.odesk.com/dpodw/serve/latest/contractor_predicted_tier/'
      },
      solr: {
        profile: 'http://172.27.68.250:9898/profiles/select',
        job: 'http://172.27.68.250:10300/openings/select'
      },
      odr: {
        profile: 'http://dev06:9000/odr/profile/f/',
        job: 'http://dev06:9000/odr/opening/f/'
      },
      ods: {
        skill: 'http://dev06:10000/ods/agg_skill_detail/pa',
        skills: 'http://dev06:10000/ods/skill/list_ex'
      },
      orpc: {
        app: 'https://admin:strange!@dev03.odesk.com:30161/orpc/',
        url : 'https://dev03.odesk.com:30161/rpcapi/index.php',
        uri   : 'http://dev03.odesk.com:30160/orpc/',
        authorization : 'Basic ZHNva29sb3Zza3k6cGFzc3cwcmQ=',
        origin  : 'https://dev03.odesk.com:30161'
      },
      db: {
        host: 'dev06.odesk.com',
        credentials: {
          'odesk_qa': {login: 'odesk_qa', password: 'odesk_qa'},
          'odesk_ba': {login: 'odesk_ba', password: 'odesk_ba'},
          'odesk_obo': {login: 'odesk_obo', password: 'odesk_obo'},
          'odesk_ua': {login: 'odesk_ua', password: 'odesk_ua'},
          'odb2': {login: 'odb2_user', password: 'odb2_user'},
          'odesk_projects': {login: 'odesk_projects', password: 'LAJYvLthyhUFaxNPJeuI'},
          'odesk_misc': {login: 'odesk_misc', password: 'odesk_misc'}
        }
      },

      subscriptionBatchServicesDSEndPoint : 'http://front.dev.agora.odesk.com:8374/current/trigger?',
      roomsEndPointsUrl: '',
      personPostEndPointsUrl: 'http://front.dev.agora.odesk.com:8224/persons',
      personEndPointsUrl: 'http://front.dev.agora.odesk.com:8064/persons',
      contactEndPointsUrl: 'http://front.dev.agora.odesk.com:8224/contacts',
      subscriptionEndPointsUrl: 'http://front.dev.agora.odesk.com:8268/subscription',
      pricingEndPointsUrl: 'http://front.dev.agora.odesk.com:8268/pricing',
      connectsEndPointsUrl: 'http://front.dev.agora.odesk.com:8282/connects',
      promotionsEndPointsUrl: 'http://front.dev.agora.odesk.com:8328'

    },
    {
      envConfig: "stash",
      skillsMapping: 'http://stash.odesk.com/projects/ATS/repos/search-utils-bundle/browse/Resources/config/skills_mapping.yml',
      topSkills: 'http://stash.odesk.com/projects/ATS/repos/profile-search-ui/browse/src/Odesk/Bundle/ProfileSearchBundle/Resources/config/top_skills.yml',
      metaSkills: 'http://stash.odesk.com/projects/ATS/repos/search-utils-bundle/browse/Resources/config/meta_skills.yml',
      metaCategories: 'http://stash.odesk.com/projects/ATS/repos/search-utils-bundle/browse/Resources/config/meta_categories.yml'
    }
  ];

  this.output = this.envDataSet.filter(function(x){
    return x.envConfig == browser.params.testenv;
  });

  this.data = this.output[0];
};

module.exports = envConf;	//naming the page object
