// protractor-core/conf/eom_messenger_conf.js
/**
 * @name nwp_conf.js
 * @description Standard conf to be use for automation.
 *
 */

var HtmlReporter = require('protractor-html-screenshot-reporter');
var baseConf = require('./base_conf.js');
var baseConfig = new baseConf();

exports.config = {

  seleniumAddress: 'http://localhost:4723/wd/hub',

  baseUrl: 'http://localhost:' + (process.env.HTTP_PORT || '8000'),

  capabilities: {
    'browserName': '',
    'appium-version': '1.0',
    'platformName': 'iOS',
    'platformVersion': '7.1',
    'app': '/Users/lle/Library/Developer/Xcode/DerivedData/oDeskMessenger-fcoerwdmzinmxwbqptvyavuqkwzr/Build/Products/Debug-iphonesimulator/oDeskMessenger.app',
    'deviceName': 'iPhone 5',
    'udid': 'udid',
    'autoAcceptAlerts': 'true'
  },

  params: {
    report: 'result',
    globalTimeout: baseConfig.timeout.pageLoad,
    waitForAngular: {
      angularFlag: true,  //Only use false if timeout/interval for polling is failing.
      angularDelay: baseConfig.timeout.angularDelay //this only goes into affect if angularFlag=false
    }
  },

  allScriptsTimeout: baseConfig.timeout.allScriptsTimeout,

  onPrepare: function () {
    browser.manage().timeouts().pageLoadTimeout(baseConfig.timeout.pageLoad);
    browser.manage().timeouts().implicitlyWait(baseConfig.timeout.implicitWait);
    browser.driver.manage().window().setSize(400,800);
    // Add a screenshot reporter and store screenshots to results folder.  It will create folder at the location protractor was ran.
    jasmine.getEnv().addReporter(new HtmlReporter({
      baseDirectory: browser.params.report,
      docTitle: 'NWP Reports',
      takeScreenShotsOnlyForFailedSpecs: true
    }));
  },

// Options to be passed to Jasmine-node.
  jasmineNodeOpts: {
    showColors: true,
    defaultTimeoutInterval: baseConfig.timeout.defaultTimeoutInterval
  }

}
;
