// protractor-core/conf/localhost_saucelab_conf.js
var HtmlReporter = require('protractor-html-screenshot-reporter');
var baseConf = require('./base_conf.js');
var baseConfig = new baseConf();
exports.config = {
  
  seleniumServerJar: null,
  seleniumPort: null,
  sauceUser: 'lamle',
  sauceKey: '12027951-5fe7-4290-ba3c-18806ee3a3d7',
  
  multiCapabilities: [
    {
      'browserName': 'chrome'
    },

  ],

  specs: [
          '../specs/0simpleSauceLabTest_spec.js'
          ],

  params: {
    globalTimeout: baseConfig.timeout.pageLoad,
    waitForAngular: {
      angularFlag: true,
      angularDelay: baseConfig.timeout.angularDelay //this only goes into affect if angularFlag=false
    }
  },

  onPrepare: function () {
    browser.manage().timeouts().pageLoadTimeout(baseConfig.timeout.pageLoad);
    browser.manage().timeouts().implicitlyWait(baseConfig.timeout.implicitWait);
    browser.driver.manage().window().setSize(1000, 900);

    jasmine.getEnv().addReporter(new HtmlReporter({
      baseDirectory: './results',
      docTitle: 'NWP Reports',
      takeScreenShotsOnlyForFailedSpecs: true
     }));
   },

// Options to be passed to Jasmine-node.
  jasmineNodeOpts: {
    showColors: true,
    defaultTimeoutInterval: baseConfig.timeout.defaultTimeoutInterval
  }

}
;
