// protractor-core/conf/localhost_mobile_conf.js
var HtmlReporter = require('protractor-html-screenshot-reporter');
var baseConf = require('./base_conf.js');
var baseConfig = new baseConf();
exports.config = {
  
  seleniumAddress: 'http://localhost:4444/wd/hub',
  
  multiCapabilities: [
    {'browserName': 'chrome',
    	'chromeOptions': {
    	    'args': [
    	      '--keep-alive-for-test',
    	      "--window-size=320,580",
    	      '--test-type',
    	      '--use-mobile-user-agent',
    	      '--user-agent="Mozilla/5.0 (iPhone; CPU iPhone OS 7_0 like Mac OS X; en-us) AppleWebKit/537.51.1 (KHTML, like Gecko) Version/7.0 Mobile/11A465 Safari/9537.53"'
    	    ]
    	  },
    	//shardTestFiles: true,
        //maxInstances: 5,
    },   
  ],

  specs: [
          '../specs/0simpleMobileTest_spec.js',
          //'../specs/WP-302_app_login_spec.js',
          ],

  params: {
    globalTimeout: baseConfig.timeout.pageLoad,
    waitForAngular: {
      angularFlag: true,
      angularDelay: baseConfig.timeout.angularDelay
    }
  },

  onPrepare: function () {
    browser.manage().timeouts().pageLoadTimeout(baseConfig.timeout.pageLoad);
    browser.manage().timeouts().implicitlyWait(baseConfig.timeout.implicitWait);
    browser.ignoreSynchronization = true; //This will tell protractor that this is a normal page or else it will keep on waiting. Our normal mobile browser is still standard.
    browser.driver.manage().window().setSize(1000, 900);
    jasmine.getEnv().addReporter(new HtmlReporter({
      baseDirectory: './results',
      docTitle: 'NWP Reports',
      takeScreenShotsOnlyForFailedSpecs: true
     }));
   },

// Options to be passed to Jasmine-node.
  jasmineNodeOpts: {
    showColors: true,
    defaultTimeoutInterval: baseConfig.timeout.defaultTimeoutInterval
  }

}
;
