// protractor-core/conf/localhost_phantomjs_conf.js
var HtmlReporter = require('protractor-html-screenshot-reporter');
var baseConf = require('./base_conf.js');
var baseConfig = new baseConf();
exports.config = {

  seleniumAddress: 'http://localhost:4444/wd/hub',

  multiCapabilities: [
    {browserName: 'phantomjs', 'phantomjs.binary.path': require('phantomjs').path}  //FYI ghostdriver is not behaving correctly right now; will look further into it at a later time.
  ],

  specs: ['../specs/WP-302_spec.js'],

  params: {
    globalTimeout: baseConfig.timeout.pageLoad,
    waitForAngular: {
      angularFlag: true,
      angularDelay: baseConfig.timeout.angularDelay
    }
  },

  onPrepare: function () {
    browser.manage().timeouts().pageLoadTimeout(baseConfig.timeout.pageLoad);
    browser.manage().timeouts().implicitlyWait(baseConfig.timeout.implicitWait);
    browser.driver.manage().window().setSize(1600, 900);
    jasmine.getEnv().addReporter(new HtmlReporter({
      baseDirectory: './results',
      docTitle: 'NWP Reports',
      takeScreenShotsOnlyForFailedSpecs: true
      }));
   },

// Options to be passed to Jasmine-node.
  jasmineNodeOpts: {
    showColors: true,
    defaultTimeoutInterval: baseConfig.timeout.defaultTimeoutInterval
  }

}
;
