// protractor-core/conf/ptc_conf.js
/**
 * Created by jso on 7/23/15.
 */
/**
 * @name ptc_conf.js
 * @description Standard conf to be use for automation.
 *
 */


var jasmine2ScreenshotReporter = require('protractor-jasmine2-screenshot-reporter');
var baseConf = require('./base_conf.js');
var baseConfig = new baseConf();

exports.config = {  //PT-Testcase_spec

  seleniumAddress: 'http://localhost:4444/wd/hub',

  specs: [
    '../specs/PT-WP-lam_spec.js'
  ],

  params: {
    report: 'result',
    globalTimeout: baseConfig.timeout.pageLoad,
    waitForAngular: {
      angularFlag: true,  //Only use false if timeout/interval for polling is failing.
      angularDelay: baseConfig.timeout.angularDelay //this only goes into affect if angularFlag=false
    }
  },

  allScriptsTimeout: baseConfig.timeout.allScriptsTimeout,

  framework: 'jasmine2',

  onPrepare: function () {
    browser.manage().timeouts().pageLoadTimeout(baseConfig.timeout.pageLoad);
    browser.manage().timeouts().implicitlyWait(baseConfig.timeout.implicitWait);
    browser.driver.manage().window().setSize(1000,1000);

    // jasmine2 screenshot reporter
    jasmine.getEnv().addReporter(
      new jasmine2ScreenshotReporter({
        dest: browser.params.report,
        filename: browser.params.reportfileName +'.html',//'Reports.html',
        reportOnlyFailedSpecs: false,
        captureOnlyFailedSpecs: true
      }));

    var jasmineReporters = require('jasmine-reporters');

    jasmine.getEnv().addReporter(
      new jasmineReporters.JUnitXmlReporter({
        savePath: browser.params.report,
        consolidateAll: false
      }));

    jasmine.getEnv().addReporter(
      new jasmineReporters.NUnitXmlReporter({
        savePath: browser.params.report
      }));

    jasmine.getEnv().addReporter(
      new jasmineReporters.TapReporter());

    jasmine.getEnv().addReporter(
      new jasmineReporters.TerminalReporter({
        verbosity: 3,
        color: true,
        showStack: true
      }));


  },

// Options to be passed to Jasmine-node.
  jasmineNodeOpts: {
    showColors: true,
    defaultTimeoutInterval: baseConfig.timeout.defaultTimeoutInterval
  }

}
;
