// protractor-core/conf/simulator_conf.js

var HtmlReporter = require('protractor-html-screenshot-reporter');
var baseConf = require('./base_conf.js');
var baseConfig = new baseConf();


exports.config = {

    //The address of a running selenium server.
    seleniumAddress: 'http://0.0.0.0:5555/wd/hub',

    //Capabilities to be passed to the webdriver instance.
    capabilities: {
        browserName: 'Safari',
        platformName: 'iOS',
        platformVersion: '7.0',
        deviceName: 'iPhone',
        device : 'iphone',
        simulator: true

    },

    //Specify the name of the specs files.
    specs: [
        '../specs/WP-808_spec.js'
    ],

    params: {
        report: 'result',
        globalTimeout: baseConfig.timeout.pageLoad,
        waitForAngular: {
            angularFlag: true,  //Only use false if timeout/interval for polling is failing.
            angularDelay: baseConfig.timeout.angularDelay //this only goes into affect if angularFlag=false
        }
    },

    allScriptsTimeout: baseConfig.timeout.allScriptsTimeout,

    onPrepare: function () {
        jasmine.getEnv().addReporter(new HtmlReporter({
            baseDirectory: browser.params.report,
            docTitle: 'Reports',
            takeScreenShotsOnlyForFailedSpecs: true
        }));
    },

// Options to be passed to Jasmine-node.
    jasmineNodeOpts: {
        showColors: true,
        defaultTimeoutInterval: baseConfig.timeout.defaultTimeoutInterval
    }

};
